src/baseConfig.js

```diff
@@ -0,0 +1,127 @@
+import extend from 'extend';
+
+/**
+ * BaseConfig class
+ */
+export class BaseConfig {
+
+  /**
+   * The defaults of your config
+   * @param {Object}
+   */
+  defaults = {};
+
+  /**
+   * The current values of your config
+   * @type {Object}
+   */
+  current = {};
+
+  /**
+   * Derived classes must set a unique alias
+   * @type {String}
+   */
+  alias
+
+  /**
+   * Creates an instance of BaseConfig. Copies defaults into current
+   * @param  {Object} defaultsOverwrite = {} Optionally extends the defaults
+   */
+  constructor(defaultsOverwrite = {}) {
+    extendIn(this, 'defaults', defaultsOverwrite);
+    this.current = JSON.parse(JSON.stringify(this.defaults));
+  }
+
+  /**
+   * Fetches a value from the current configuration
+   * @param  {string|Array<keys>} keys (dot-separated) string(s) or array of keys
+   * @return {any}                that The retrived value
+   */
+  fetch(...keys) {
+    return fetchFrom(this.current, ...normalizeKey(...keys));
+  }
+
+  /**
+   * Deletes a value from the current configuration
+   * @param {string|Array<keys>} keys (dot-separated) string(s) or array of keys
+   */
+  delete(...keys) {
+    delete this.fetch(...keys);
+  }
+
+  /**
+   * Sets a value in the current configuration
+   * @param {string|Array<keys>} keys  (dot-separated) string(s) or array of keys
+   * @param {any}                value The value to set
+   */
+  set(keys, value) {
```

Comment:
no. the ...rest is only allowed to be the last. could be done with using arguments i guess